File: Project Netwerken/src/NodeJS/global/model/command/IDRequest.js
const Command = require('./Command');
const varint = require('../../utils/varint/varint');

class IDRequest extends Command {
    constructor(options) {
        super(options);
        if (options.buffer) {
            // constructor(buffer)
            this.random = varint.decode(options.buffer);
        } else {
            // constructor(nodeId, vectorClock)
            this.random = Math.floor(Math.random() * 4228250625 + 1)
        }
    }

/*    // OLD CONSTRUCTOR
    constructor(buffer) {
        super(buffer);
        this.random = varint.decode(buffer);
    }

    // OLD CONSTRUCTOR
    constructor(nodeId, vectorClock) {
        super(1, nodeId, vectorClock);
        this.random = Math.floor(Math.random() * 4228250625 + 1);
    }*/

    write(buffer) {
        super.write(buffer);
        buffer.write(varint.encode(this.random));
    }
}

module.exports = IDRequest;

File: Project Netwerken/src/Global/src/model/command/IDRequest.java
package model.command;

import utils.VarInt;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;

/**
 * request vectorclock index/nodeId from server.
 */
public class IDRequest extends Command {
    private static int TYPE = 1;
    private long random = -1; //temp id

    public IDRequest(ByteBuffer buffer) {
        super(TYPE, buffer);
        random = VarInt.readUnsignedVarLong(buffer);
    }

    /**
     * request vectorclock index from server.
     */
    public IDRequest(VectorClock vectorClock) {
        super(TYPE, 0, vectorClock);
        this.random = getRandom();
    }

    /**
     * @return temp id
     */
    public long getRandom() {
        if(random == -1){
            random = (long) (Math.random() * Long.MAX_VALUE + 1);
            assert random >= 0;
        }
        return random;
    }

    @Override
    public void write(ByteArrayOutputStream out) throws IOException {
        super.write(out);
        VarInt.writeUnsignedVarLong(random, out);
    }
}
